### PR TITLE
Fix some issues with `powershell -Command -`

### LINKED ISSUE #2071: (PS 5.1.14393) Passing commands via STDIN in Ruby does not seem to work
Steps to reproduce
------------------
1. Install Ruby 2.3.0

2. Create a file in the root of the repository `test.rb` with the following content
```ruby
require 'open3'
require 'timeout'

cmd = 'powershell.exe -NoProfile -NonInteractive -NoLogo -ExecutionPolicy Bypass -Command -'
str1 = <<-CODE


Write-Output "1"
Write-Output "2"
Write-Output "3"


CODE
# Helper
def drain_pipe(pipe, iterations = 10)
  output = []
  0.upto(iterations) do
    break if !is_readable?(pipe, 0.1)
    l = pipe.gets
    puts "#{Time.now} PIPE> #{l}"
    output << l
  end
  output
end
def self.is_readable?(stream, timeout = 0.5)
  read_ready = IO.select([stream], [], [], timeout)
  read_ready && stream == read_ready[0][0]
end


## Let's go
puts "Starting PS..."
@stdin, @stdout, @stderr, @ps_process = Open3.popen3(cmd)

puts "Enter to send STDIN"
blah = gets

puts "Sending STDIN..."
@stdin.puts(str1)
sleep(2)
output = drain_pipe(@stdout)

puts "output = #{output}"

puts "Enter to cleanup"
blah = gets

#  CLEANUP
puts "Cleaning up..."
@stdin.puts "\nexit\n"
@stdin.close
@stdout.close
@stderr.close

begin
Timeout.timeout(3) do
  puts "Awaiting PowerShell process termination..."
  @exit_status = @ps_process.value
end
rescue Timeout::Error
end

exit_msg = "PowerShell process exited: #{@exit_status}" if @exit_status
puts(exit_msg)
if @ps_process.alive?
  puts("Forcefully terminating PowerShell process.")
  Process.kill('KILL', @ps_process[:pid])
end
```

3. run 
```
ruby test.rb
```

4. Press enter when prompted

Expected behavior
-----------------
```
Starting PS...
Enter to send STDIN

Sending STDIN...
2016-08-25 16:10:24 -0700 PIPE> 1
2016-08-25 16:10:24 -0700 PIPE> 2
2016-08-25 16:10:24 -0700 PIPE> 3
output = ["1\n", "2\n", "3\n"]
Enter to cleanup

Cleaning up...
Awaiting PowerShell process termination...
PowerShell process exited: pid 2788 exit 0
```

Actual behavior
---------------
```
Starting PS...
Enter to send STDIN

Sending STDIN...
output = []
Enter to cleanup

Cleaning up...
Awaiting PowerShell process termination...

Forcefully terminating PowerShell process.
```

Environment data
----------------
Doesn't work
```
Windows 10 Enterprise


Name                           Value
----                           -----
PSVersion                      5.1.14393.82
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14393.82
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

Does Work
```
Server 2012 R2

Name                           Value
----                           -----
PSVersion                      5.0.10586.117
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.10586.117
CLRVersion                     4.0.30319.34209
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

```

Workaround
----------------
Modify the initial powershell command from
`cmd = 'powershell.exe -NoProfile -NonInteractive -NoLogo -ExecutionPolicy Bypass -Command -'`

to

`cmd = 'powershell.exe -Version 2.0 -NoProfile -NonInteractive -NoLogo -ExecutionPolicy Bypass -Command -'`

This works on the Windows 10 system (PS 5.1)

### PR DESCRIPTION
**Pull request by lzybkr**
Fixes #2071 and #2319.

### CODE REVIEW

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs [hunk 1]
```diff
@@ -2868,6 +2868,25 @@ private class ConsoleHostStartupException : Exception
...
+
+#if CORECLR
+        internal Lazy<TextReader> ConsoleIn { get; } = new Lazy<TextReader>(() => Console.In);
+#else
+        internal Lazy<TextReader> ConsoleIn { get; } = new Lazy<TextReader>(() =>
+        {
+            // This is a workaround for a full clr issue that causes a hang when calling PowerShell from ruby.
+            // They use named pipes instead of anonymous pipes, and for some reason that triggers a hang
+            // reading from Console.In.
+            var inputHandle = ConsoleControl.NativeMethods.GetStdHandle(-10);
```
Comment (vors) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: What this `-10` mean? Can you add comment about it?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs [hunk 2]
```diff
@@ -1588,7 +1591,9 @@ private string ReadLineFromFile(string initialContent)
                     break;
                 }
 
-                if (c == '\b')
+                // If NoPrompt is true, we are in a sort of server mode where we shouldn't
+                // do anything like edit the command line - every character is part of the input.
+                if (c == '\b' && !NoPrompt)
```
Comment (vors) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs]: Not sure about that. IO redirection scenario could be used to marshal interactive session as well, even with NoPrompt. I.e. in something like EditorServices. Did this caused mentioned bugs or it's just for tests?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 1 above)
Comment (lzybkr) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: Is it not obvious that it's standard in? (std handle, assigned to var named inputHandle, and comment mentions Console.In)

**Review by vors — APPROVED** [review]

**Review by PaulHigin — APPROVED** [review]
LGTM

**Comment by ARBFL** [comment]
When is this fix available for Windows 1607? Will there be KB / fix in Windows Updates?

**Comment by iSazonov** [comment]
@ARBFL This PR is closed.  You would be better to create Issue-Question.

**Pull request merged**